elf: skip BenchmarkWriteELF if ELF file wasn't built

tklauser opened the pull request:
When running `go test -run=^$ -bench=. ./...` on the Cilium source,
BenchmarkWriteELF currently fails with:

      --- FAIL: BenchmarkWriteELF
          elf_test.go:201: failed to open ELF file ../../test/bpf/elf-demo.o: open ../../test/bpf/elf-demo.o: no such file or directory

That particular file is only built when running tests/benchmarks using
the Makefile-based flow. In order to still allow Cilium's benchmarks
suite to run without errors when invoked using native `go test -bench`,
just skip BenchmarkWriteELF if the ELF file doesn't exist.

In the future we might want to embed that ELF file into the
test/benchmark itself using bpf2go [1].

[1] https://pkg.go.dev/github.com/cilium/ebpf/cmd/bpf2go

For #17535

odeke-em commented on pkg/elf/elf_test.go:
```diff
@@ -197,7 +198,9 @@ func BenchmarkWriteELF(b *testing.B) {
 
 	elf, err := Open(baseObjPath)
-	if err != nil {
+	if errors.Is(err, os.ErrNotExist) {
```
The recommendation as you know is to use fs.ErrNotExist but I guess to support more versions of Go you are using os.ErrNotExist. I think if we are in this direction we might as well just do os.IsNotExist(err) to handle any syscall edges too?

odeke-em replied:
It would be nice to also add a comment referencing the issue for posterity:

// If the ELF file couldn't be found most likely it
// wasn't built. See issue \<link>

tklauser replied:
@odeke-em thanks for the feedback. If initially tried `os.IsNotExist` but that didn't seem to work. Guess we're hitting https://golang.org/issues/38198. I don't see a recommendation anywhere to prefer fs.ErrNotExist over os.ErrNotExist in the godoc or the Go release notes. Where does it say that the former is preferred?

I'll add that comment referencing the issue, good point.

odeke-em replied:
It is recommendEd at https://pkg.go.dev/os#IsNotExist
![image](https://user-images.githubusercontent.com/4898263/136174262-ee67e93e-599f-45ab-bebd-f30557ad7f48.jpeg)
Plus I was part of the reviews for the fs package hence the preference for it but what you have works, thank you!

odeke-em reviewed (approved):
LGTM, thank you @tklauser!

tklauser committed:
elf: skip BenchmarkWriteELF if ELF file wasn't built

When running `go test -run=^$ -bench=. ./...` on the Cilium source,
BenchmarkWriteELF currently fails with:

  --- FAIL: BenchmarkWriteELF
        elf_test.go:201: failed to open ELF file ../../test/bpf/elf-demo.o: open ../../test/bpf/elf-demo.o: no such file or directory

That particular file is only built when running tests/benchmarks using
the Makefile-based flow. In order to still allow Cilium's benchmarks
suite to run without errors when invoked using native `go test -bench`,
just skip BenchmarkWriteELF if the ELF file doesn't exist.

In the future we might want to embed that ELF file into the
test/benchmark itself using bpf2go [1].

[1] https://pkg.go.dev/github.com/cilium/ebpf/cmd/bpf2go

For #17535

Reported-by: Emmanuel T Odeke <emmanuel@orijtech.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

odeke-em commented on pkg/elf/elf_test.go:
```diff
@@ -197,7 +198,12 @@ func BenchmarkWriteELF(b *testing.B) {
 
 	elf, err := Open(baseObjPath)
-	if err != nil {
+	b.Logf("%#v", err)
```
Did you mean to leave this log in?

tklauser replied:
:facepalm: nope, that wasn't intended. I also switched to `fs.ErrNotExist` per the recommondation. Thanks @odeke-em!

odeke-em reviewed (approved):
LGTM++ with one comment. Thank you @tklauser!

odeke-em reviewed (approved):
LGTM++ 👏🏿🤙🏾🎉 let's ship it, thank you @tklauser!
